src/drivers/magnetometer/isentek/ist8310/IST8310.cpp

```diff
@@ -0,0 +1,298 @@
+		break;
+
+	case STATE::WAIT_FOR_RESET:
+
+		// SRST: This bit is automatically reset to zero after POR routine
+		if ((RegisterRead(Register::WAI) == Device_ID)
+		    && ((RegisterRead(Register::CNTL2) & CNTL2_BIT::SRST) == 0)) {
+
+			// if reset succeeded then configure
+			_state = STATE::CONFIGURE;
+			ScheduleDelayed(10_ms);
+
+		} else {
+			// RESET not complete
+			if (hrt_elapsed_time(&_reset_timestamp) > 1000_ms) {
+				PX4_DEBUG("Reset failed, retrying");
+				_state = STATE::RESET;
+				ScheduleDelayed(100_ms);
+
+			} else {
+				PX4_DEBUG("Reset not complete, check again in 10 ms");
+				ScheduleDelayed(10_ms);
+			}
+		}
+
+		break;
+
+	case STATE::CONFIGURE:
+		if (Configure()) {
+			// if configure succeeded then start measurement cycle
+			_state = STATE::MEASURE;
+			ScheduleDelayed(20_ms);
+
+		} else {
+			// CONFIGURE not complete
+			if (hrt_elapsed_time(&_reset_timestamp) > 1000_ms) {
+				PX4_DEBUG("Configure failed, resetting");
+				_state = STATE::RESET;
+
+			} else {
+				PX4_DEBUG("Configure failed, retrying");
+			}
+
+			ScheduleDelayed(100_ms);
+		}
+
+		break;
+
+	case STATE::MEASURE:
+		RegisterWrite(Register::CNTL1, CNTL1_BIT::MODE_SINGLE_MEASUREMENT);
+		_state = STATE::READ;
+		ScheduleDelayed(20_ms); // Wait at least 6ms. (minimum waiting time for 16 times internal average setup)
+		break;
+
+	case STATE::READ: {
+			struct TransferBuffer {
+				uint8_t STAT1;
+				uint8_t DATAXL;
+				uint8_t DATAXH;
+				uint8_t DATAYL;
+				uint8_t DATAYH;
+				uint8_t DATAZL;
+				uint8_t DATAZH;
+			} buffer{};
+
+			bool success = false;
+			uint8_t cmd = static_cast<uint8_t>(Register::STAT1);
+
+			if (transfer(&cmd, 1, (uint8_t *)&buffer, sizeof(buffer)) == PX4_OK) {
+
+				if (buffer.STAT1 && STAT1_BIT::DRDY) {
+					int16_t x = combine(buffer.DATAXH, buffer.DATAXL);
+					int16_t y = combine(buffer.DATAYH, buffer.DATAYL);
+					int16_t z = combine(buffer.DATAZH, buffer.DATAZL);
+
+					// sensor's frame is +x forward, +y right, +z up
+					z = (z == INT16_MIN) ? INT16_MAX : -z; // flip z
+
+					_px4_mag.update(now, x, y, z);
+
+					success = true;
+
+					if (_failure_count > 0) {
+						_failure_count--;
+					}
+				}
+
+			} else {
+				perf_count(_bad_transfer_perf);
+			}
+
+			if (!success) {
+				_failure_count++;
+
+				// full reset if things are failing consistently
+				if (_failure_count > 10) {
+					Reset();
+					return;
+				}
+			}
+
+			if (!success || hrt_elapsed_time(&_last_config_check_timestamp) > 100_ms) {
+				// check configuration registers periodically or immediately following any failure
+				if (RegisterCheck(_register_cfg[_checked_register])) {
+					_last_config_check_timestamp = now;
+					_checked_register = (_checked_register + 1) % size_register_cfg;
+
+				} else {
+					// register check failed, force reset
+					perf_count(_bad_register_perf);
+					Reset();
+					return;
+				}
+			}
+
+			// initiate next measurement
+			_state = STATE::MEASURE;
+			ScheduleNow();
```

Comment:
Just writing the register would be a bit more efficient here.

Reply:
I suppose it's a bit much. Fixed.